flink-examples/flink-examples-streaming/src/main/java/org/apache/flink/streaming/examples/async/AsyncIOExample.java

```diff
@@ -116,8 +116,8 @@ public void cancel() {
 	private static class SampleAsyncFunction extends RichAsyncFunction<Integer, String> {
 		private static final long serialVersionUID = 2098635244857937717L;
 
-		private static ExecutorService executorService;
-		private static Random random;
```

Comment:
with this change, I think line 148 is no longer needed

Reply:
right, line 118, not line 148

Reply:
No, my point is: if we are making executorService and random instance-specific - then we don't need to synchronize on class (lines 148 and 163).
If we want to keep executorService and random as static, then we need to make counter static too (line 122). 
I think having all three (executorService, random, counter) as static is best.

Having executorService and random as static, but counter as instance-specific will not work when someone creates a second instance of SampleAsyncFunction. In second SampleAsyncFunction the counter will be 0 and we will re-intialize static executorService and random, thus interfering with the first SampleAsyncFunction object.

Reply:
SampleAsyncFunction is used as singleton, so synchronize is needed

Reply:
What makes it a singleton?.. 
Nothing stops me from adding a call in main():
`AsyncFunction<Integer, String> function2 = new SampleAsyncFunction(sleepFactor2, failRatio2, shutdownWaitTS2);`

IMHO, we need to put back "static" on executorService and random. And add "static" to counter.